### PR TITLE
[Mapping] Implicit IDAttribute

### PR DESCRIPTION
**Pull request by Lewiky**
Adds the implicit ID Attribute to every service block, including adding `id` to the list of reserved keywords in attribute names. 

Most of the changes here are in tests, which I must say are so much nicer to work with now! Thanks @jaylees14

### CODE REVIEW

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 1]
```diff
@@ -56,6 +56,7 @@ object NameClashes {
     "grant",
     "group",
     "having",
+    "id",
```
Comment (c-harding) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala]: ID has been added elsewhere, this is just a list of Postgres reserved words

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 2]
```diff
@@ -37,7 +37,8 @@ private class Validator private (templefile: Templefile) {
     // Keep a set of names that have been used already
     val takenNames: mutable.Set[String] = attributes.keys.to(mutable.Set)
 
-    attributes.map {
+    // Add implicit ID Attribute
```
Comment (c-harding) [inline src/main/scala/temple/DSL/semantics/Validator.scala:41]: I thought I was getting Scala `implicit`s 😢

File: src/main/scala/temple/builder/DatabaseBuilder.scala [hunk 3]
```diff
@@ -24,10 +24,15 @@ object DatabaseBuilder {
   private def toColDef(name: String, attribute: Attribute): ColumnDef = {
     val nonNullConstraint = when(!attribute.valueAnnotations.contains(Nullable)) { ColumnConstraint.NonNull }
 
+    val primaryKeyConstraint = attribute match {
+      case IDAttribute => Some(ColumnConstraint.PrimaryKey)
+      case _           => None
+    }
```
Comment (c-harding) [inline src/main/scala/temple/builder/DatabaseBuilder.scala]: inclined to suggest `when(attribute == IDAttribute) { ColumnConstraint.PrimaryKey }`, but up to you

File: src/test/resources/project-builder-complex/complex-user/dao/dao.go [hunk 4]
```diff
@@ -112,7 +112,7 @@ func executeQuery(db *sql.DB, query string, args ...interface{}) (int64, error)
...
+	row := executeQueryWithRowResponse(dao.DB, "INSERT INTO complex_user (id, smallIntField, intField, bigIntField, floatField, doubleField, stringField, boundedStringField, boolField, dateField, timeField, dateTimeField, blobField) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) RETURNING id, smallIntField, intField, bigIntField, floatField, doubleField, stringField, boundedStringField, boolField, dateField, timeField, dateTimeField, blobField;", input.ID, input.SmallIntField, input.IntField, input.BigIntField, input.FloatField, input.DoubleField, input.StringField, input.BoundedStringField, input.BoolField, input.DateField, input.TimeField, input.DateTimeField, input.BlobField)
```
Comment (c-harding) [inline src/test/resources/project-builder-complex/complex-user/dao/dao.go:115]: 🤢

**Review by c-harding — CHANGES_REQUESTED** [review]
nice, only thing that really needs fixing is the validation of ID (that list is cited, so shouldn’t be updated)

**Review by c-harding — APPROVED** [review]
Has SQL’s ILIKE been duplicated?

File: src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go [hunk 5]
```diff
@@ -77,7 +77,7 @@ func executeQuery(db *sql.DB, query string, args ...interface{}) (int64, error)
 
 // CreateSimpleTempleTestGroup creates a new simpleTempleTestGroup in the datastore, returning the newly created simpleTempleTestGroup
 func (dao *DAO) CreateSimpleTempleTestGroup(input CreateSimpleTempleTestGroupInput) (*SimpleTempleTestGroup, error) {
-	row := executeQueryWithRowResponse(dao.DB, "INSERT INTO simple_temple_test_group () VALUES ();", input.ID, input.AuthID)
+	row := executeQueryWithRowResponse(dao.DB, "INSERT INTO simple_temple_test_group (id) VALUES ($1) RETURNING id;", input.ID, input.AuthID)
```
Comment (jaylees14) [inline src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go:80]: I see here there's also `input.AuthID` passed in (i.e. a `$2`) - is that going to be used?

File: src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go [hunk 6]
```diff
@@ -108,7 +108,7 @@ func (dao *DAO) ReadSimpleTempleTestGroup(input ReadSimpleTempleTestGroupInput)
 
 // UpdateSimpleTempleTestGroup updates the simpleTempleTestGroup in the datastore for a given ID, returning the newly updated simpleTempleTestGroup
 func (dao *DAO) UpdateSimpleTempleTestGroup(input UpdateSimpleTempleTestGroupInput) (*SimpleTempleTestGroup, error) {
-	row := executeQueryWithRowResponse(dao.DB, "UPDATE simple_temple_test_group SET WHERE id = $1;", input.ID)
+	row := executeQueryWithRowResponse(dao.DB, "UPDATE simple_temple_test_group SET WHERE id = $1 RETURNING id;", input.ID)
 
 	var simpleTempleTestGroup SimpleTempleTestGroup
 	err := row.Scan(&simpleTempleTestGroup.ID, &simpleTempleTestGroup.CreatedBy)
```
Comment (jaylees14) [inline src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go:114]: This also looks a little broken (I know it wasn't your change - but should we be storing that too?)

@smithwjv is this on your radar?

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 7]
```diff
@@ -175,6 +175,7 @@ object NameClashes {
     "group",
     "having",
     "ilike", // sql, it’s my favourite language
+    "ilike", // sql, it’s my favourite language
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala]: So good you added it twice? `yy p`? 😉

**Comment by Lewiky** [comment]
> Has SQL’s ILIKE been duplicated?

Yup - fixed now

**Review by jaylees14 — APPROVED** [review]
Good stuff, nice and easy to follow 👍

File: src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go (hunk 6 above)
Comment (willjvsmith) [reply src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go:114]: Yeah aha, made a card a few days ago: "Handle the case of no user-defined/client-facing attributes", pesky one

File: src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go (hunk 5 above)
Comment (willjvsmith) [reply src/e2e/resources/simple-temple-expected/simple-temple-test-group/dao/dao.go:80]: That's my problem to fix, project not having any auth isn't handled by DAO generation - also made a card for it a few days ago 😉

**Pull request merged**